Remove `token::{Open,Close}Delim`

nnethercote opened the pull request:
By replacing them with `{Open,Close}{Param,Brace,Bracket,Invisible}`.

PR #137902 made `ast::TokenKind` more like `lexer::TokenKind` by
replacing the compound `BinOp{,Eq}(BinOpToken)` variants with fieldless
variants `Plus`, `Minus`, `Star`, etc. This commit does a similar thing
with delimiters. It also makes `ast::TokenKind` more similar to
`parser::TokenType`.

This requires a few new methods:
- `TokenKind::is_{,open_,close_}delim()` replace various kinds of
  pattern matches.
- `Delimiter::as_{open,close}_token_kind` are used to convert
  `Delimiter` values to `TokenKind`.

Despite these additions, it's a net reduction in lines of code. This is
because e.g. `token::OpenParen` is so much shorter than
`token::OpenDelim(Delimiter::Parenthesis)` that many multi-line forms
reduce to single line forms. And many places where the number of lines
doesn't change are still easier to read, just because the names are
shorter, e.g.:
```
-   } else if self.token != token::CloseDelim(Delimiter::Brace) {
+   } else if self.token != token::CloseBrace {
```

r? @petrochenkov

nnethercote committed:
Remove `token::{Open,Close}Delim`.

By replacing them with `{Open,Close}{Param,Brace,Bracket,Invisible}`.

PR #137902 made `ast::TokenKind` more like `lexer::TokenKind` by
replacing the compound `BinOp{,Eq}(BinOpToken)` variants with fieldless
variants `Plus`, `Minus`, `Star`, etc. This commit does a similar thing
with delimiters. It also makes `ast::TokenKind` more similar to
`parser::TokenType`.

This requires a few new methods:
- `TokenKind::is_{,open_,close_}delim()` replace various kinds of
  pattern matches.
- `Delimiter::as_{open,close}_token_kind` are used to convert
  `Delimiter` values to `TokenKind`.

Despite these additions, it's a net reduction in lines of code. This is
because e.g. `token::OpenParen` is so much shorter than
`token::OpenDelim(Delimiter::Parenthesis)` that many multi-line forms
reduce to single line forms. And many places where the number of lines
doesn't change are still easier to read, just because the names are
shorter, e.g.:
```
-   } else if self.token != token::CloseDelim(Delimiter::Brace) {
+   } else if self.token != token::CloseBrace {
```

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14487417731/job/40635855295) [(plain)](https://github.com/rust-lang/rust/commit/f6f6a74f53e658b5bfd6ab50996633f8c668f432/checks/40635855295/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Mismatch at src/parse/macros/cfg_if.rs:60:
             return Err("Expected an opening brace");
         }
 
-        while parser.token != TokenKind::CloseBrace
-            && parser.token.kind != TokenKind::Eof
-        {
+        while parser.token != TokenKind::CloseBrace && parser.token.kind != TokenKind::Eof {
             let item = match parser.parse_item(ForceCollect::No) {
                 Ok(Some(item_ptr)) => item_ptr.into_inner(),
                 Ok(None) => continue,
{ "type": "test", "name": "test::self_tests", "event": "failed", "stdout": "Ran 8 self tests.\n\nthread 'test::self_tests' panicked at src/tools/rustfmt/src/test/mod.rs:420:5:\nassertion `left == right` failed: 1 self tests failed\n  left: 1\n right: 0\nnote: run with `RUST_BACKTRACE=1` environment variable to display a backtrace\n" }
test test::system_tests ... ok
test test::idempotence_tests ... ok

test result: FAILED. 181 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 1.88s

```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-19`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14488469847/job/40639213165) [(plain)](https://github.com/rust-lang/rust/commit/8318b85f750f84d2f7f5ba7436ebc5161d8b3501/checks/40639213165/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#19 exporting to docker image format
#19 sending tarball 19.7s done
#19 DONE 26.3s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-19]
[CI_JOB_NAME=x86_64-gnu-llvm-19]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-19', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-19/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Number of decisions:   4447
  longest path:          1159 (code:    152)
  longest backtrack:       66 (code:    428)
Shared 86733 out of 152951 states by creating 14756 new states, saving 71977
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/expmed.cc: In function ‘rtx_def* extract_bit_field_1(rtx, poly_uint64, poly_uint64, int, rtx, machine_mode, machine_mode, bool, bool, rtx_def**)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/expmed.cc:1864:45: warning: ‘*(unsigned int*)((char*)&imode + offsetof(scalar_int_mode, scalar_int_mode::m_mode))’ may be used uninitialized [-Wmaybe-uninitialized]
 1864 |       rtx sub = extract_bit_field_as_subreg (mode1, op0, imode,
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
 1865 |                                              bitsize, bitnum);
      |                                              ~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/expmed.cc:1824:19: note: ‘*(unsigned int*)((char*)&imode + offsetof(scalar_int_mode, scalar_int_mode::m_mode))’ was declared here
 1824 |   scalar_int_mode imode;
      |                   ^~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gimple-range-gori.cc: In member function ‘void range_def_chain::dump(FILE*, basic_block, const char*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gimple-range-gori.cc:319:19: warning: format not a string literal and no format arguments [-Wformat-security]
  319 |           fprintf (f, prefix);
      |           ~~~~~~~~^~~~~~~~~~~
---
                 from /checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/analyzer/region-model.h:33,
                 from /checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/analyzer/access-diagram.cc:39:
In constructor ‘ana::byte_range::byte_range(ana::byte_offset_t, ana::byte_size_t)’,
    inlined from ‘virtual text_art::table ana::string_literal_spatial_item::make_table(const ana::bit_to_table_map&, text_art::style_manager&) const’ at /checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/analyzer/access-diagram.cc:1812:18:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/analyzer/store.h:312:5: warning: ‘size_in_bytes.generic_wide_int<fixed_wide_int_storage<128> >::fixed_wide_int_storage<128>.fixed_wide_int_storage<128>::val[1]’ may be used uninitialized [-Wmaybe-uninitialized]
  312 |     m_size_in_bytes (size_in_bytes)
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/analyzer/access-diagram.cc: In member function ‘virtual text_art::table ana::string_literal_spatial_item::make_table(const ana::bit_to_table_map&, text_art::style_manager&) const’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/analyzer/access-diagram.cc:1808:28: note: ‘size_in_bytes.generic_wide_int<fixed_wide_int_storage<128> >::fixed_wide_int_storage<128>.fixed_wide_int_storage<128>::val[1]’ was declared here
 1808 |                byte_size_t size_in_bytes
      |                            ^~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/diagnostic.cc: In function ‘void fancy_abort(const char*, int, const char*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/diagnostic.cc:1677:15: warning: format not a string literal and no format arguments [-Wformat-security]
 1677 |       fnotice (stderr, diagnostic_kind_text[DK_ICE]);
---
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gcc.cc:7930:9: warning: ignoring return value of ‘ssize_t write(int, const void*, size_t)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 7930 |   write (fd, "\n\n", 2);
      |   ~~~~~~^~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gcc.cc: In member function ‘void driver::final_actions() const’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gcc.cc:9307:13: warning: ignoring return value of ‘int truncate(const char*, __off_t)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 9307 |     truncate(totruncate_file, 0);
      |     ~~~~~~~~^~~~~~~~~~~~~~~~~~~~
In file included from /checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-playback.h:31,
                 from /checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-builtins.cc:24:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-recording.h: In member function ‘virtual bool gcc::jit::recording::type::is_same_type_as(gcc::jit::recording::type*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-recording.h:640:20: warning: suggest parentheses around ‘&&’ within ‘||’ [-Wparentheses]
---
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc: In function ‘void lto_resolution_read(splay_tree, FILE*, lto_file*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc:2091:10: warning: ignoring return value of ‘int fscanf(FILE*, const char*, ...)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 2091 |   fscanf (resolution, " ");   /* Read white space.  */
      |   ~~~~~~~^~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc:2093:9: warning: ignoring return value of ‘size_t fread(void*, size_t, size_t, FILE*)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 2093 |   fread (obj_name, sizeof (char), name_len, resolution);
      |   ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc:2113:10: warning: ignoring return value of ‘int fscanf(FILE*, const char*, ...)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 2113 |   fscanf (resolution, "%u", &num_symbols);
      |   ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-recording.cc:32:
---
Applying io_quotes_use            to linux/blkzoned.h
Applying io_quotes_use            to linux/ipmi.h
Applying io_quotes_use            to linux/psp-dbc.h
Applying io_quotes_use            to linux/bt-bmc.h
Applying io_quotes_use            to linux/tps6594_pfsm.h
Applying io_quotes_use            to linux/cxl_mem.h
Applying io_quotes_use            to linux/wmi.h
Applying io_quotes_use            to linux/auto_fs.h
Applying io_quotes_use            to linux/mmtimer.h
Applying io_quotes_use            to linux/f2fs.h
Applying io_quotes_use            to linux/vhost.h
---
Applying io_quotes_use            to sound/asound.h
Applying io_quotes_use            to sound/compress_offload.h
Applying hpux8_bogus_inlines      to math.h
Applying pthread_incomplete_struct_argument to pthread.h
Fixed:  pthread.h
Applying io_quotes_use            to misc/mrvl_cn10k_dpi.h
Applying io_quotes_use            to misc/ocxl.h
Applying io_quotes_use            to misc/cxl.h
Applying io_quotes_use            to misc/xilinx_sdfec.h
Applying io_quotes_def            to unicode/platform.h
Applying sun_malloc               to malloc.h
Applying io_quotes_use            to scsi/cxlflash_ioctl.h
---
Applying machine_name             to x86_64-linux-gnu/bits/unistd_ext.h
Applying io_quotes_use            to x86_64-linux-gnu/asm/mtrr.h
Applying io_quotes_use            to x86_64-linux-gnu/asm/amd_hsmp.h
Applying machine_name             to openssl/e_os2.h
Applying io_quotes_use            to drm/xe_drm.h
Applying io_quotes_use            to drm/radeon_drm.h
Applying io_quotes_use            to drm/panfrost_drm.h
Applying io_quotes_use            to drm/etnaviv_drm.h
Applying io_quotes_use            to drm/lima_drm.h
Applying io_quotes_use            to drm/qaic_accel.h
Applying io_quotes_use            to drm/vc4_drm.h
Applying io_quotes_use            to drm/i915_drm.h
Applying io_quotes_use            to drm/omap_drm.h
Applying io_quotes_use            to drm/pvr_drm.h
Applying io_quotes_use            to drm/amdgpu_drm.h
Applying io_quotes_use            to drm/vgem_drm.h
Applying io_quotes_use            to drm/msm_drm.h
Applying io_quotes_use            to drm/v3d_drm.h
Applying io_quotes_use            to drm/exynos_drm.h
Applying io_quotes_use            to drm/nouveau_drm.h
Applying io_quotes_use            to drm/drm.h
Applying io_quotes_use            to drm/habanalabs_accel.h
Applying io_quotes_use            to drm/tegra_drm.h
Applying io_quotes_use            to rdma/rdma_user_ioctl.h
cc1: note: self-tests are not enabled in this build
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/c++tools/server.cc: In function ‘void server(bool, int, module_resolver*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/c++tools/server.cc:620:10: warning: ignoring return value of ‘int pipe(int*)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
---
---- [rustdoc] tests/rustdoc/macro-generated-macro.rs stdout ----

error: htmldocck failed!
status: exit status: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/macro-generated-macro" "/checkout/tests/rustdoc/macro-generated-macro.rs"
--- stdout -------------------------------
--- expected ---

macro_rules! morestuff {
    (
        <= "space between most kinds of tokens" : 1 $x + @ :: >>= 'static
        "no space inside paren or bracket" : (2 a) [2 a] $(2 $a:tt)*
        "space inside curly brace" : { 2 a }
        "no space inside empty delimiters" : () [] {}
        "no space before comma or semicolon" : a, (a), { a }, a; [T; 0];
        "the three repetition specifiers" : $(@)*, $(@)+, $(@)?
        "repetition separators" : $(@)|*, $(@)|+, $(@)==*, $(@)static*
        "plus or star cannot be a repetition separator" : $(@)+ * $(@)* +
        "no space between ident and paren" : let _ = f(0) + f[0] + Struct {};
        "space between keyword and paren" : return (a,) & for x in (..)
        "some special case keywords" : pub(crate), fn() -> u8, Self(0, 0) =>
    ) => { ... };
}


--- actual ---

macro_rules! morestuff {
    (
        <= "space between most kinds of tokens" : 1 $x + @ :: >>= 'static
        "no space inside paren or bracket" : (2 a( [2 a[ $(2 $a:tt(*
        "space inside curly brace" : { 2 a {
        "no space inside empty delimiters" : (( [[ {{
        "no space before comma or semicolon" : a, (a(, { a {, a; [T; 0[;
        "the three repetition specifiers" : $(@(*, $(@(+, $(@(?
        "repetition separators" : $(@(|*, $(@(|+, $(@(==*, $(@(static*
        "plus or star cannot be a repetition separator" : $(@(+ * $(@(* +
        "no space between ident and paren" : let _ = f(0( + f[0[ + Struct {{;
        "space between keyword and paren" : return (a,( & for x in (..(
        "some special case keywords" : pub(crate(, fn(( -> u8, Self(0, 0( =>
    ) => { ... };
}
------------------------------------------
--- stderr -------------------------------
11: has check failed
 `XPATH PATTERN` did not match
 //@ has - //pre '(<= type $($i:ident)::* + $e:expr =>) => { ... };'
16: has check failed
 `XPATH PATTERN` did not match
 //@ has - //pre '(<= #![no_std] #[cfg(feature = "alloc")] =>) => { ... };'
20: has check failed
 `XPATH PATTERN` did not match
 //@ has - //pre '(<= fn {} () { foo[0] } =>) => { ... };'
26: snapshot check failed
 Actual snapshot value is different than expected
 //@ snapshot macro_morestuff_pre macro_generated_macro/macro.morestuff.html //pre/text()

Encountered 4 errors
------------------------------------------


```

</details>

nnethercote commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit d6622d6345bcfb120dfd747db4f58cccc2317171 with merge 448bb7314ad82fe33016e188f27ebe57376d93cd...
<!-- homu: {"type":"TryBuildStarted","head_sha":"d6622d6345bcfb120dfd747db4f58cccc2317171","merge_sha":"448bb7314ad82fe33016e188f27ebe57376d93cd"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/14491454509/job/40656572752)
Build commit: 448bb7314ad82fe33016e188f27ebe57376d93cd (`448bb7314ad82fe33016e188f27ebe57376d93cd`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/14491454509/job/40656572752"},"merge_sha":"448bb7314ad82fe33016e188f27ebe57376d93cd"} -->

rust-timer commented:
Queued 448bb7314ad82fe33016e188f27ebe57376d93cd with parent afa859f8121bf2985362a2c8414dc71a825ccf2d, future [comparison URL](https://perf.rust-lang.org/compare.html?start=afa859f8121bf2985362a2c8414dc71a825ccf2d&end=448bb7314ad82fe33016e188f27ebe57376d93cd).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.4 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (448bb7314ad82fe33016e188f27ebe57376d93cd): [comparison URL](https://perf.rust-lang.org/compare.html?start=afa859f8121bf2985362a2c8414dc71a825ccf2d&end=448bb7314ad82fe33016e188f27ebe57376d93cd&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=afa859f8121bf2985362a2c8414dc71a825ccf2d&end=448bb7314ad82fe33016e188f27ebe57376d93cd&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.4% | [-0.5%, -0.2%] | 14    |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.4% | [-0.5%, -0.2%] | 14    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=afa859f8121bf2985362a2c8414dc71a825ccf2d&end=448bb7314ad82fe33016e188f27ebe57376d93cd&stat=max-rss)
<details>
<summary>Results (primary -0.4%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.8%  | [2.8%, 2.8%]   | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.7% | [-1.1%, -0.4%] | 10    |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.4% | [-1.1%, 2.8%]  | 11    |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=afa859f8121bf2985362a2c8414dc71a825ccf2d&end=448bb7314ad82fe33016e188f27ebe57376d93cd&stat=cycles:u)
<details>
<summary>Results (primary -0.6%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.6% | [-0.7%, -0.4%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.6% | [-0.7%, -0.4%] | 2     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=afa859f8121bf2985362a2c8414dc71a825ccf2d&end=448bb7314ad82fe33016e188f27ebe57376d93cd&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 776.859s -> 775.51s (-0.17%)
**Artifact size**: 364.72 MiB -> 364.80 MiB (0.02%)

nnethercote commented:
Slight perf wins too, that's nice.

petrochenkov commented on compiler/rustc_parse/src/parser/diagnostics.rs:
```diff
@@ -1470,9 +1462,7 @@ impl<'a> Parser<'a> {
-                        if !&[token::OpenDelim(Delimiter::Parenthesis), token::PathSep]
-                            .contains(&self.token.kind)
-                        {
+                        if !&[token::OpenParen, token::PathSep].contains(&self.token.kind) {
```
```suggestion
                        if !matches!(self.token.kind, token::OpenParen | token::PathSep) {
```

petrochenkov commented on compiler/rustc_parse/src/parser/item.rs:
```diff
@@ -399,14 +399,9 @@ impl<'a> Parser<'a> {
-                    token::OpenDelim(Delimiter::Parenthesis),
-                ]
-                .contains(&t.kind)
+                [token::Lt, token::OpenBrace, token::OpenParen].contains(&t.kind)
```
Also can be `matches!`.

petrochenkov commented on compiler/rustc_parse/src/parser/stmt.rs:
```diff
@@ -874,8 +868,7 @@ impl<'a> Parser<'a> {
                     && !expr.attrs.is_empty()
-                    && ![token::Eof, token::Semi, token::CloseDelim(Delimiter::Brace)]
-                        .contains(&self.token.kind) =>
+                    && ![token::Eof, token::Semi, token::CloseBrace].contains(&self.token.kind) =>
```
Also `matches!`.

petrochenkov commented:
Overall looks reasonable.
@rustbot author

nnethercote commented:
I added two new commits.
- The first addresses the review comments. I grepped for `matches!` calls that could be converted to `==`, and found a few more.
- The second removes an extraneous `diff` file that was accidentally added in #139904.

nnethercote commented:
@rustbot ready

jieyouxu commented:
Can you drop the diff removal commit (since there's #139991 already)?

nnethercote commented:
> Can you drop the `diff` removal commit (since there's #139991 already)?

Done.

nnethercote commented:
I squashed the commits.

@bors r=petrochenkov

bors commented:
:pushpin: Commit bf8ce32558a4657d077a6761eaa293d0645c2e16 has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov bf8ce32558a4657d077a6761eaa293d0645c2e16 -->
<!-- homu: {"type":"Approved","sha":"bf8ce32558a4657d077a6761eaa293d0645c2e16","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit bf8ce32558a4657d077a6761eaa293d0645c2e16 with merge fae7785b60ea7fe1ad293352c057a5b7be73d245...
<!-- homu: {"type":"BuildStarted","head_sha":"bf8ce32558a4657d077a6761eaa293d0645c2e16","merge_sha":"fae7785b60ea7fe1ad293352c057a5b7be73d245"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/14584547445/job/40913823933)
Approved by: petrochenkov
Pushing fae7785b60ea7fe1ad293352c057a5b7be73d245 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/14584547445/job/40913823933"},"merge_sha":"fae7785b60ea7fe1ad293352c057a5b7be73d245"} -->

rust-timer commented:
Finished benchmarking commit (fae7785b60ea7fe1ad293352c057a5b7be73d245): [comparison URL](https://perf.rust-lang.org/compare.html?start=d6c1e454aa8af5e7e59fbf5c4e7d3128d2f99582&end=fae7785b60ea7fe1ad293352c057a5b7be73d245&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=d6c1e454aa8af5e7e59fbf5c4e7d3128d2f99582&end=fae7785b60ea7fe1ad293352c057a5b7be73d245&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.4% | [-0.6%, -0.2%] | 14    |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.4% | [-0.6%, -0.2%] | 14    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=d6c1e454aa8af5e7e59fbf5c4e7d3128d2f99582&end=fae7785b60ea7fe1ad293352c057a5b7be73d245&stat=max-rss)
<details>
<summary>Results (primary 0.3%, secondary 3.4%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.7%  | [0.5%, 1.0%]   | 10    |
| Regressions ❌ <br /> (secondary)  | 3.4%  | [3.4%, 3.4%]   | 1     |
| Improvements ✅ <br /> (primary)   | -1.8% | [-2.0%, -1.5%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.3%  | [-2.0%, 1.0%]  | 12    |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=d6c1e454aa8af5e7e59fbf5c4e7d3128d2f99582&end=fae7785b60ea7fe1ad293352c057a5b7be73d245&stat=cycles:u)
<details>
<summary>Results (primary -0.6%, secondary 1.8%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 2.4%  | [2.1%, 3.0%]   | 6     |
| Improvements ✅ <br /> (primary)   | -0.6% | [-0.6%, -0.5%] | 2     |
| Improvements ✅ <br /> (secondary) | -1.6% | [-1.6%, -1.6%] | 1     |
| All ❌✅ (primary)                 | -0.6% | [-0.6%, -0.5%] | 2     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=d6c1e454aa8af5e7e59fbf5c4e7d3128d2f99582&end=fae7785b60ea7fe1ad293352c057a5b7be73d245&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 774.01s -> 775.257s (0.16%)
**Artifact size**: 365.04 MiB -> 365.06 MiB (0.01%)
